operators/katib-controller/src/charm.py

```diff
@@ -0,0 +1,309 @@
+                            },
+                            {
+                                "name": "metrics",
+                                "containerPort": self.model.config["metrics-port"],
+                            },
+                        ],
+                        "envConfig": {
+                            "KATIB_CORE_NAMESPACE": self.model.name,
+                        },
+                        "volumeConfig": [
+                            {
+                                "name": "certs",
+                                "mountPath": "/tmp/cert",
+                                "files": [
+                                    {
+                                        "path": "tls.crt",
+                                        "content": self._stored.cert,
+                                    },
+                                    {
+                                        "path": "tls.key",
+                                        "content": self._stored.key,
+                                    },
+                                ],
+                            }
+                        ],
+                    }
+                ],
+            },
+            k8s_resources={
+                "kubernetesResources": {
+                    "customResourceDefinitions": [
+                        {"name": crd["metadata"]["name"], "spec": crd["spec"]}
+                        for crd in yaml.safe_load_all(Path("src/crds.yaml").read_text())
+                    ],
+                    "mutatingWebhookConfigurations": [
+                        {
+                            "name": mutating["metadata"]["name"],
+                            "webhooks": mutating["webhooks"],
+                        }
+                    ],
+                    "validatingWebhookConfigurations": [
+                        {
+                            "name": validating["metadata"]["name"],
+                            "webhooks": validating["webhooks"],
+                        }
+                    ],
+                },
+                "configMaps": {
+                    "katib-config": {
+                        f: Path(f"src/{f}.json").read_text()
+                        for f in (
+                            "metrics-collector-sidecar",
+                            "suggestion",
+                            "early-stopping",
+                        )
+                    },
+                    "trial-template": {
+                        f + suffix: Path(f"src/{f}.yaml").read_text()
+                        for f, suffix in (
+                            ("defaultTrialTemplate", ".yaml"),
+                            ("enasCPUTemplate", ""),
+                            ("pytorchJobTemplate", ""),
+                        )
+                    },
+                },
+            },
+        )
+
+        self.model.unit.status = ActiveStatus()
+
+    def gen_certs(self):
```

Comment:
Why do you manually generate cert in your script if we have `cert-generator` now ?

Reply:
This is kind of the crux of the operator framework that these charms are written in. Instead of relying on building special images to handle the extra work that can't be expressed well in straight Kubernetes YAML, there's one general purpose operator pod that can handle a variety of tasks related to setting up the workload pod.

Reply:
Got it, thank you!